src/common/src/cache.rs

```diff
@@ -769,23 +770,45 @@ impl<K: LruKey, T: LruValue> LruCache<K, T> {
 
 pub struct CleanCacheGuard<'a, K: LruKey + Clone + 'static, T: LruValue + 'static> {
     cache: &'a Arc<LruCache<K, T>>,
-    key: K,
+    key: Option<K>,
     hash: u64,
-    success: bool,
+}
+
+impl<'a, K: LruKey + Clone + 'static, T: LruValue + 'static> CleanCacheGuard<'a, K, T> {
+    fn mark_success(&mut self) -> K {
+        self.key.take().unwrap()
+    }
 }
 
 impl<'a, K: LruKey + Clone + 'static, T: LruValue + 'static> Drop for CleanCacheGuard<'a, K, T> {
     fn drop(&mut self) {
-        if !self.success {
-            self.cache.clear_pending_request(&self.key, self.hash);
+        if let Some(key) = self.key.as_ref() {
+            self.cache.clear_pending_request(key, self.hash);
         }
     }
 }
 
-/// Only implement `lookup_with_request_dedup` for static values, as they can be sent across tokio
-/// spawned futures.
+/// `lookup_with_request_dedup` can directly return `Result<CacheableEntry<K, T>, E>`, but if we do
+/// not want to wait when cache miss happens, we can call `lookup_with_request_dedup_raw` which will
+/// return `JoinHandle<Result<CacheableEntry<K, T>, E>>` when cache miss happens.
+pub enum LookupResponse<K: LruKey + Clone + 'static, T: LruValue + 'static, E> {
+    Cached(CacheableEntry<K, T>),
+    Miss(JoinHandle<Result<CacheableEntry<K, T>, E>>),
```

Comment:
Better to rename it to `Pending`. 

And we can return a `BoxFuture<'static, Result<CacheableEntry<K, T>, E>>` instead.

Reply:
> BoxFuture<'static, Result<CacheableEntry<K, T>, E>>

fixed